File: ECanopy/ECanopy/Models/AmenityBooking.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

//Soumya Prasad Sahu - Part-G Amenities Booking

namespace ECanopy.Models
{
    public class AmenityBooking
    {
        [Key]
        public int AmenityBookingId { get; set; }

        public int AmenityId { get; set; }

        [ForeignKey(nameof(AmenityId))]
        public Amenity Amenity { get; set; } = null!;

        public int ResidentId { get; set; }

        [ForeignKey(nameof(ResidentId))]
        public Resident Resident { get; set; } = null!;

        public DateTime BookingDate { get; set; }

        public TimeSpan StartTime { get; set; }

        public TimeSpan EndTime { get; set; }

        public int NumberOfGuests { get; set; }

        public string Status { get; set; } = "Pending";

        public DateTime RequestedAt { get; set; } = DateTime.UtcNow;

        public DateTime? ApprovedAt { get; set; }

        public string? ApprovedByUserId { get; set; }

        [ForeignKey(nameof(ApprovedByUserId))]
        public ApplicationUser? ApprovedByUser { get; set; }

        public string? Notes { get; set; }
    }
}

File: backend/src/main/java/com/ecanopy/entity/AmenityBooking.java
package com.ecanopy.entity;

import com.ecanopy.entity.enums.BookingStatus;
import jakarta.persistence.*;
import jakarta.validation.constraints.Min;
import lombok.*;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;

@Entity
@Table(name = "amenity_bookings", uniqueConstraints = {
        @UniqueConstraint(columnNames = { "amenity_id", "booking_date", "start_time" })
})
@EntityListeners(AuditingEntityListener.class)
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class AmenityBooking {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long amenityBookingId;

    @Column(nullable = false)
    private LocalDate bookingDate;

    @Column(nullable = false)
    private LocalTime startTime;

    @Column(nullable = false)
    private LocalTime endTime;

    @Min(0)
    private Integer numberOfGuests;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    @Builder.Default
    private BookingStatus status = BookingStatus.PENDING;

    @CreatedDate
    @Column(nullable = false, updatable = false)
    private LocalDateTime requestedAt;

    private LocalDateTime approvedAt;

    @Column(length = 500)
    private String notes;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "amenity_id", nullable = false)
    private Amenity amenity;

    @com.fasterxml.jackson.annotation.JsonIgnore
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "resident_id", nullable = false)
    private Resident resident;

    @com.fasterxml.jackson.annotation.JsonIgnore
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "approved_by_user_id")
    private User approvedByUser;
}
